docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,230 @@
+
+### `types` API Reference
+
+This section details the initial programming interface we propose for `type`s when they are reflected into type function bodies. Each section is separated by headers (e.g. Singleton) and describe the methods available to that category of type. The initial section, titled `types`, describes the elements of the interface common to every category of type. All attributes of newly-created `type`s are initialized with empty tables / arrays and `nil`. All arguments are passed by references. 
+
+<details><summary>Expand for full type API reference.</summary>
+
+#### `types`
+
+| Instance Attributes | Type | Description |
+| ------------- | ------------- | ------------- |
+| `unknown` | `type` | an immutable instance of the built-in type `unknown` |
+| `never` | `type` | an immutable instance of the built-in type `never` |
+| `any` | `type` | an immutable instance of the built-in type `any` |
+| `boolean` | `type` | an immutable instance of the built-in type `boolean` |
+| `number` | `type` | an immutable instance of the built-in type `number` |
+| `string` | `type` | an immutable instance of the built-in type `string` |
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `__eq(arg: type)` | `boolean` | overrides the == operator to return true if self is syntactically equal to arg |
+| `is(arg: string)` | `boolean` | returns true if self has the same tag as the argument. List of available tags: "nil", "unknown", "never", "any", "boolean", "number", "string", "singleton", "negation", "union", "intersection", "table", "function", "class"  |
+
+| Static Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `singleton(arg: string \| boolean \| nil)` | `type` | returns an immutable instance of the argument as a singleton type |
+| `negationof(arg: type)` | `type` | returns an immutable instance of the parameter as a negated type; the parameter cannot be a table or function type |
+| `unionof(...: type)` | `type` | returns an immutable instance of an union type of the arguments; requires at least 2 parameters to be provided |
+| `intersectionof(...: type)` | `type` | returns an immutable instance of an intersection type of the arguments; requires at least 2 parameters to be provided |
+| `newtable(props: {[type]: type}?, indexer: {key: type, value: type}?, metatable: type?)` | `type` | returns a mutable instance of a table type; the keys of the table's property must be a singleton type; if provided the metatable parameter, this table becomes a metatable |
```

Comment:
There is no 'metatable' tag in type function context. If we really want to explain the internals here (on how runtime value is converted into typechecker type), it would be best to say `MetatableType`.
But some can say that separate representation chosen for tables with metatable in the typechecker wasn't the best choice :)